### DISCUSSION TITLE
login with username

### DISCUSSION THREAD

**Discussion by paul45** [codeigniter4/shield#323] [Q&A] [answered]
I try to use validFields to login with username
the login views only display a email field with email validation

Is this function operational?

**Comment by datamweb** [comment, marked as answer]
Yes, you can.
1. Edit file `app/Config/Auth.php   `        
```php 
    public array $views = [
        'login'                       => '\App\Views\Shield\login',
    ];
```
and comment `email ` and uncomment `username`:
```php
    /**
     * --------------------------------------------------------------------
     * Valid login fields
     * --------------------------------------------------------------------
     * Fields that are available to be used as credentials for login.
     */
    public array $validFields = [
        //'email',
        'username',
    ];
```
2. copy file `vendor\codeigniter4\shield\src\Views\login.php` to `app\Views\Shield\login.php`  and edit :
```html
<!-- Email -->
<div class="mb-2">
     <input type="text" class="form-control" name="username" placeholder="<?= lang('Auth.usernam') ?>" value="<?= old('username') ?>" required />
</div>
```
3. Edit file `vendor\codeigniter4\shield\src\Controllers\LoginController.php `
```php
    protected function getValidationRules(): array
    {
        return setting('Validation.login') ?? [
            'username' => config('AuthSession')->usernameValidationRules,
           // 'email'    => config('AuthSession')->emailValidationRules,
            'password' => 'required',
        ];
    }
```
I have not tested but it should work.
In the 3 step, it is **better to create a custom login controller** (the explanation given is **not the right way**, it is only for summary)

**Reply by paul45** [reply]
thx i will try that

**Reply by paul45** [reply]
yes it works fine

**Reply by kenjis** [reply]
See https://shield.codeigniter.com/customization/validation_rules/
This is Registration, but you can customize Login validation rules in the same way.

**Comment by vago-me** [comment]
Is it possible to log in with a username or email using the same input box? like GitHub.

**Reply by datamweb** [reply]
Shield ensures you can only use one of those valid fields as a credential during login with the Session controller.
